clippy_lints/src/duration_to_float_precision_loss.rs

```diff
@@ -0,0 +1,409 @@
+                "as_micros" => DurationMethod::AsMicros,
+                "as_nanos" => DurationMethod::AsNanos,
+                _ => {
+                    // does not match criteria - not the type of duration methods we're interested in
+                    return None;
+                },
+            };
+
+            Some(LintApplicableSite {
+                duration_method,
+                precision,
+                expr_span: expr.span,
+                adjustment: None,
+                duration_expr_span: method_receiver_expr.span,
+            })
+        } else {
+            // does not match criteria - method receiver type is not Duration
+            None
+        }
+    } else {
+        // does not match criteria - not a method call expression in the expression of the cast
+        None
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+enum FloatPrecision {
+    F32,
+    F64,
+}
+
+impl FloatPrecision {
+    fn as_secs_float_method(self) -> &'static str {
+        match self {
+            FloatPrecision::F32 => "as_secs_f32",
+            FloatPrecision::F64 => "as_secs_f64",
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+#[allow(clippy::enum_variant_names)]
+enum DurationMethod {
+    AsSeconds,
+    AsMillis,
+    AsMicros,
+    AsNanos,
+}
+
+#[derive(Debug, Copy, Clone)]
+enum AdjustmentScale {
+    Thousand,
+    Million,
+    Billion,
+}
+
+#[derive(Debug, Copy, Clone)]
+enum AdjustmentDirection {
+    /// Decrease the number - multiply by factor < 1, or divide by factor > 1
+    Negative,
+    /// Increase the number - multiply by factor > 1, or divide by factor < 1
+    Positive,
+}
+
+impl AdjustmentDirection {
+    fn negate(self) -> Self {
+        match self {
+            AdjustmentDirection::Positive => AdjustmentDirection::Negative,
+            AdjustmentDirection::Negative => AdjustmentDirection::Positive,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct LintApplicableSite {
+    precision: FloatPrecision,
+    expr_span: Span,
+    duration_expr_span: Span,
+    duration_method: DurationMethod,
+    // When the adjustment is missing from the site, set this to None
+    adjustment: Option<(AdjustmentDirection, AdjustmentScale)>,
+}
+
+impl LintApplicableSite {
+    /// Return a string with the suggestion or None if this lint is not applicable.
+    #[allow(clippy::enum_glob_use)]
+    fn suggestion(&self) -> Option<String> {
+        use AdjustmentScale::*;
+
+        let (method_name, scale) = apply_adjustment(self.precision, self.duration_method, self.adjustment)?;
+        let scale = match scale {
+            None => "",
+            Some(Thousand) => " * 1e3",
+            Some(Million) => " * 1e6",
+            Some(Billion) => " * 1e9",
+        };
+
+        let suggestion = format!("{method_name}(){scale}");
+        Some(suggestion)
+    }
+
+    fn emit_lint(&self, cx: &LateContext<'_>) {
+        let mut applicability = Applicability::MachineApplicable;
```

Comment:
Would be good to use `MaybeIncorrect` since there's a behaviour change. A note making it clear that the suggestion would include fractional e.g. milliseconds would also be good